GraphQL-Schema-Design.md

```diff
@@ -0,0 +1,142 @@
+## Notes
+
+* GraphQL exposes a data set as [a graph](https://en.wikipedia.org/wiki/Graph_theory). See
+  [Explaining GraphQL Connections](https://dev-blog.apollodata.com/explaining-graphql-connections-c48b7c3d6976) for a
+  more detailed overview on how this applies to our codebase.
+
+# How to model our graph
+
+* As with most GraphQL advice, our schema should strive to always be backwards compatible.
+  Thus careful consideration is needed about naming (be as unambiguous as possible) and the shape of the data returned
+  by a field.
+
+* Whenever a ‘namespace’ would be used in a field by adding a prefix or suffix (e.g. `for_sale_artwork` or
+  `location_city`), this should be taken as a clue that this data needs to be nested instead. E.g.
+
+  ```graphql
+  {
+    show(id: "kate-oh-gallery-metropolis") {
+      # Don’t do this:
+      location_city
+      location_country
+      # Instead do this:
+      location {
+        city
+        country
+      }
+    }
+   }
+  ```
+
+* Design the schema around first class domain-models, not functional details.
+
+  For instance, rather than mimicking a backend endpoint that allows one to filter artworks by defining a
+  `filter_artworks` field that has a nested artworks connection, expose the ability to filter artworks in a plain
+  `artworks` connection field instead.
+
+  The key aspect here is that we are trying to expose **artworks**, which are the same as artworks retrieved through
+  other means, only the ‘feed’ they are retrieved from is different and that’s an implementation detail, there is no
+  such model as **filter artworks**.
+
+  ```graphql
+  # Bad
+  {
+    filter_artworks(aggregations: [TOTAL]) {
+      counts {
+        total
+      }
+      artworks(first: 10) {
+        edges {
+          node {
+            title
+          }
+        }
+      }
+    }
+  }
+
+  # Good
+  {
+    artworks(aggregations: [TOTAL], first: 10) {
+      counts {
+        total
+      }
+      edges {
+        node {
+          title
+        }
+      }
+    }
+  }
+  ```
+
+## Root fields
+
+* These are entry points into the graph that is Artsy’s data set.
+
+* Not all types necessarily make sense as root fields.
+  - In general only domain models should be exposed as root fields. An ‘artwork’ is such a model, Artsy’s system time
```

Comment:
We should strife to add as few root fields as possible in order to keep the schema browsable. For Artsy, root fields like `artwork` and `sale` make sense, but `system_time` isn’t a first class model, just a functional detail. I’d rather see such a field modelled like:

```graphql
{
  system {
    time
  }
}
```

so that there’s a single root field that all other such function details can be grouped in.

On reflection, the documentation here is definitely lacking to convey that, so clarification would be much appreciated!

Reply:
Yeah, I can see that. Although, I also see the risk in shoehorning a bunch of unrelated things into grab bags like `system`. Feels like a `/lib` directory :). Not having spent a huge amount of time browsing schemas yet, I haven't run into challenges with field proliferation, so I'll take your word for it!

But where I do see that problem popping up much more aggressively is in mutations, given that in REST, there are several mutation verbs for the one query verb. Is namespacing mutations seen as a good idea, or should they just go right on the mutation type?

Reply:
Yeah you’re totally right (I usually refer to it as the ‘utils’ bag but pretty sure we mean the same thing). Maybe the example here is the wrong one, when @ashfurrow PR’d a system time field to MP, I actually suggested it would go in e.g. the sale type, and all other places where a need to sync is required.

With regards to mutations I don’t have enormous gah experience of my own yet and also haven’t noticed much discussion about it in the community. It’s a good point and we should look into that more sooner rather than later.

Reply:
Yup, the PR I made is here if it helps: https://github.com/artsy/metaphysics/pull/812